### PR TITLE
diem contract

### PR DESCRIPTION
**Pull request by goldenfiredo**

### CODE REVIEW

File: pruntime/enclave/src/contracts/diem.rs [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+                                }
+                            }
+                        }
+
+                        if !found {
+                            println!("bad sender/receiver address");
+                            return Ok(Response::VerifyTransaction { total, verified: false });
+                        }
+                    } else {
+                        if let Some(seq) = self.seq_number.get(&account_address) {
```
Comment (h4x3rotab) [inline pruntime/enclave/src/contracts/diem.rs:173]: Let's add a comment here: "// Outgoing tx must be synced sequencely"

File: pruntime/enclave/src/contracts/diem.rs [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+                        Ok(tx) => tx,
+                        Err(_) => {
+                            println!("decode transaction error");
+                            return Ok(Response::VerifyTransaction { total, verified: false });
+                        }
+                    };
+
+                    let signed_tx: SignedTransaction = transaction.as_signed_user_txn().unwrap().clone();
+                    if signed_tx.raw_txn.sender != address {
+                        let mut found = false;
```
Comment (h4x3rotab) [inline pruntime/enclave/src/contracts/diem.rs:156]: Add comment: "// Incoming tx doesn't need to be sequential"

File: pruntime/enclave/src/contracts/diem.rs [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+                                        found = true;
+                                        break;
+                                    }
+                                }
+                            }
+                        }
+
+                        if !found {
+                            println!("bad sender/receiver address");
+                            return Ok(Response::VerifyTransaction { total, verified: false });
```
Comment (h4x3rotab) [inline pruntime/enclave/src/contracts/diem.rs:170]: Better to define these errors and return as an error

File: pruntime/enclave/src/contracts/diem.rs [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+                    println!("account_info:{:?}", account_info);
+                    let exist = self.accounts.iter().any(|x| x.address == account_info.address);
+                    if !exist {
+                        self.accounts.push(account_info);
+                    }
+
+                    Ok(Response::AccountData { size: self.accounts.len() as u32 })
+                }
+
+                Request::VerifyTransaction {account_address, transaction_with_proof_b64} => {
```
Comment (h4x3rotab) [inline pruntime/enclave/src/contracts/diem.rs:128]: Let's move this to another function, and maybe break it to a few functions? It's tooo long

File: pruntime/enclave/src/contracts/diem.rs [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+        TransactionStatus::Ok
+    }
+
+    fn handle_query(&mut self, origin: Option<&chain::AccountId>, req: Request) -> Response {
+        let inner = || -> Result<Response, Error> {
+            match req {
+                Request::AccountData { account_data_b64} => {
+                    println!("account_data_b64: {:?}", account_data_b64);
+                    let account_data = base64::decode(&account_data_b64).unwrap();
+                    let account_info: AccountInfo = bcs::from_bytes(&account_data).unwrap();
```
[the remark this answers is not in the document]
Comment (goldenfiredo) [reply pruntime/enclave/src/contracts/diem.rs]: Done

File: pruntime/enclave/src/contracts/diem.rs (hunk 1 above)
Comment (goldenfiredo) [reply pruntime/enclave/src/contracts/diem.rs]: Done

File: pruntime/enclave/src/contracts/diem.rs (hunk 2 above)
Comment (goldenfiredo) [reply pruntime/enclave/src/contracts/diem.rs]: Done

File: pruntime/enclave/src/contracts/diem.rs (hunk 3 above)
Comment (goldenfiredo) [reply pruntime/enclave/src/contracts/diem.rs]: Done

File: pruntime/enclave/src/contracts/diem.rs (hunk 4 above)
Comment (goldenfiredo) [reply pruntime/enclave/src/contracts/diem.rs:264]: Done

File: pruntime/enclave/src/contracts/diem.rs [hunk 6]
```diff
@@ -100,6 +100,140 @@ impl Diem {
...
+        {
+            println!("Transaction was verified");
+
+            verified = true;
+        } else {
+            println!("Failed to verify transaction");
+        }
+
+        verified
+    }
```
Comment (h4x3rotab) [inline pruntime/enclave/src/contracts/diem.rs:236]:
```suggestion
        if let Ok(_) = account_transaction_state_proof.verify(
            ledger_info_with_signatures.ledger_info(),
            transaction_with_proof.version, address.hash(),
            Some(&transaction_with_proof.account_state_blob))
        {
            println!("Transaction was verified");
            true
        } else {
            println!("Failed to verify transaction");
            false
        }
    }
```

File: pruntime/enclave/src/contracts/diem.rs [hunk 7]
```diff
@@ -132,53 +268,22 @@ impl contracts::Contract<Command, Request, Response> for Diem {
...
 
-                    let transaction: Transaction = match bcs::from_bytes(&transaction_with_proof.transaction_bytes) {
+                    let transaction = match self.get_transaction(transaction_with_proof.clone(), account_address.clone(), address) {
                         Ok(tx) => tx,
-                        Err(_) => {
-                            println!("decode transaction error");
-                            return Ok(Response::VerifyTransaction { total, verified: false });
-                        }
+                        Err(err) => return Err(err),
                     };
```
Comment (h4x3rotab) [inline pruntime/enclave/src/contracts/diem.rs]: let transaction = self.get_transaction(transaction_with_proof.clone(), account_address.clone(), address)?;

File: pruntime/enclave/src/contracts/diem.rs [hunk 8]
```diff
@@ -196,65 +301,11 @@ impl contracts::Contract<Command, Request, Response> for Diem {
...
-                    let mut verified = false;
-                    if let Ok(_) = account_transaction_state_proof.verify(ledger_info_with_signatures.ledger_info(),
-                        transaction_with_proof.version, address.hash(), Some(&transaction_with_proof.account_state_blob)) {
-                        println!("Transaction was verified");
-
-                        verified = true;
-                    } else {
-                        println!("Transaction verification was failed");
+                    if let Err(err) = self.verify_trusted_state(transaction_with_proof.clone()) {
+                        return Err(err);
```
Comment (h4x3rotab) [inline pruntime/enclave/src/contracts/diem.rs:305]:
```suggestion
                    self.verify_trusted_state(transaction_with_proof.clone()?;
```

File: pruntime/enclave/src/contracts/diem.rs [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+
+    pub fn verify_trusted_state(
+        &mut self,
+        transaction_with_proof: TransactionWithProof,
+    ) -> Result<(), Error> {
+        let epoch_change_proof: EpochChangeProof = transaction_with_proof.epoch_change_proof.clone();
+        let ledger_info_with_signatures: LedgerInfoWithSignatures = transaction_with_proof.ledger_info_with_signatures.clone();
+        let zero_ledger_info_with_sigs = epoch_change_proof.ledger_info_with_sigs[0].clone();
+        let mut trusted_state = TrustedState::try_from(zero_ledger_info_with_sigs.ledger_info())
+            .map_err(|_|Error::Other("get TrustedState error".to_string()))?;
```
Comment (h4x3rotab) [inline pruntime/enclave/src/contracts/diem.rs:166]: We should store `trusted_state` inside `struct Diem` contract, and update it by `new_state` we got in this function

File: pruntime/enclave/src/contracts/diem.rs [hunk 10]
```diff
@@ -246,6 +251,20 @@ impl contracts::Contract<Command, Request, Response> for Diem {
     fn handle_query(&mut self, origin: Option<&chain::AccountId>, req: Request) -> Response {
         let inner = || -> Result<Response, Error> {
             match req {
+                Request::SetTrustedState { trusted_state_b64 } => {
```
Comment (h4x3rotab) [inline pruntime/enclave/src/contracts/diem.rs:254]: For the convenience of debugging, we can allow setting `trusted_state` multiple times, but it shouldn't be the case in the production. Let's add a comment here:

```
// TODO: disable in read product
```

**Pull request merged**